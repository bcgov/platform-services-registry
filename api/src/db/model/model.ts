// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-28.
//

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { transformKeysToCamelCase } from '../utils';

export interface Query {
  text: string;
  values?: any[];
}

export interface CommonFields {
  id?: number;
  archived?: boolean;
  createdAt?: object;
  updatedAt?: object;
}

export abstract class Model {
  abstract table: string;
  abstract requiredFields: string[];
  abstract pool: Pool;

  abstract async create(data: any): Promise<any>;
  abstract async update(profileId: number, data: any): Promise<any>;
  abstract async delete(profileId: number): Promise<any>;

  async findAll(): Promise<any[]> {
    const query = {
      text: `
      SELECT * FROM ${this.table}
        WHERE archived = false;
      `,
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to fetch all Profiles`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findById(id: number): Promise<any> {
    const query = {
      text: `
        SELECT * FROM ${this.table}
          WHERE id = $1 AND archived = false;`,
      values: [id],
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to fetch Profile with ID ${id}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findQuota(): Promise<any[]> {
    const query = {
      text: `
      SELECT * FROM ${this.table};
      `,
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to fetch all Quota`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async runQuery(query: Query): Promise<any[]> {
    let client;

    if (this.pool.waitingCount > 0) {
      logger.warn(`Database pool has ${this.pool.waitingCount} waiting queries`);
    }

    try {
      client = await this.pool.connect();
      const results = await client.query(query);

      return results.rows.map(transformKeysToCamelCase);
    } catch (err) {
      throw err;
    } finally {
      client.release();
    }
  }
}
