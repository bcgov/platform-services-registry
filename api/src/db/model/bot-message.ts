// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export interface BotMessage extends CommonFields {
    requestId: number;
    natsSubject: string;
    natsContext: string;
    clusterName: string;
    hasCallback: boolean;
}

export default class BotMessageModel extends Model {
    table: string = 'bot_message';
    requiredFields: string[] = [
        'request_id', 'nats_subject', 'nats_context', 'cluster_name', 'has_callback'
    ];
    pool: Pool;

    constructor(pool: any) {
        super();
        this.pool = pool;
    }

    async create(data: BotMessage): Promise<BotMessage> {
        const query = {
            text: `INSERT INTO ${this.table}
            (request_id, nats_subject, nats_context, cluster_name, has_callback)
            VALUES ($1, $2, $3, $4, $5) RETURNING *;`,
            values: [
                data.requestId,
                data.natsSubject,
                data.natsContext,
                data.clusterName,
                data.hasCallback,
            ],
        };

        try {
            const results = await this.runQuery(query);
            return results.pop();
        } catch (err) {
            const message = `Unable to create request`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }

    async update(botMessageId: number, data: BotMessage): Promise<BotMessage> {
        const values: any[] = [];
        const query = {
            text: `UPDATE ${this.table}
            SET
            request_id = $1, nats_subject = $2, nats_context = $3, cluster_name = $4, has_callback = $5
            WHERE id = ${botMessageId}
            RETURNING *;`,
            values,
        };

        try {
            const record = await this.findById(botMessageId);
            const aData = { ...record, ...data };
            query.values = [
                aData.requestId,
                aData.natsSubject,
                aData.natsContext,
                aData.clusterName,
                aData.hasCallback,
            ];

            const results = await this.runQuery(query);
            return results.pop();
        } catch (err) {
            const message = `Unable to update request ID ${botMessageId}`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }

    async delete(botMessageId: number): Promise<BotMessage> {
        const query = {
            text: `UPDATE ${this.table}
            SET
            archived = true
            WHERE id = ${botMessageId}
            RETURNING *;
        `,
        };

        try {
            const results = await this.runQuery(query);
            return results.pop();
        } catch (err) {
            const message = `Unable to archive request`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }

    async findForRequest(requestId: number): Promise<BotMessage[]> {
        const query = {
            text: `
                SELECT * FROM ${this.table}
                    WHERE request_id = ${requestId} AND archived = false;
            `,
        };

        try {
            return await this.runQuery(query);
        } catch (err) {
            const message = `Unable to fetch Request(s) with Request Id ${requestId}`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }
}
