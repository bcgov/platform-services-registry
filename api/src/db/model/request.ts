// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export interface Request extends CommonFields {
    profileId: number,
    editType: string,
    editObject: string,
    natsSubject?: string,
    natsContext?: string,
}

export default class RequestModel extends Model {
    table: string = 'request';
    requiredFields: string[] = [
        'profileId', 'editType', 'editObject'
    ];
    pool: Pool;

    constructor(pool: any) {
        super();
        this.pool = pool;
    }

    async create(data: Request): Promise<Request> {
        const query = {
            text: `INSERT INTO ${this.table}
            (profile_id, edit_type, edit_object, nats_subject, nats_context)
            VALUES ($1, $2, $3, $4, $5) RETURNING *;`,
            values: [
                data.profileId,
                data.editType,
                data.editObject,
                data.natsSubject,
                data.natsContext
            ],
        };

        try {
            const results = await this.runQuery(query);
            return results.pop();
        } catch (err) {
            const message = `Unable to create request`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }

    async update(requestId: number, data: Request): Promise<Request> {
        const values: any[] = [];
        const query = {
            text: `UPDATE ${this.table}
            SET
            profile_id = $1, edit_type = $2, edit_object = $3, nats_subject = $4, nats_context = $5
            WHERE id = ${requestId}
            RETURNING *;`,
            values,
        };

        try {
            const record = await this.findById(requestId);
            const aData = { ...record, ...data };
            query.values = [
                aData.profileId,
                aData.editType,
                aData.editObject,
                aData.natsSubject,
                aData.natsContext
            ];

            const results = await this.runQuery(query);
            return results.pop();
        } catch (err) {
            const message = `Unable to update request ID ${requestId}`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    };

    async delete(requestId: number): Promise<Request> {
        const query = {
            text: `UPDATE ${this.table}
            SET
            archived = true
            WHERE id = ${requestId}
            RETURNING *;
        `,
        };

        try {
            const results = await this.runQuery(query);
            return results.pop();
        } catch (err) {
            const message = `Unable to archive request`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    };

    async findForProfile(profileId: number): Promise<Request[]> {
        const query = {
            text: `
                SELECT * FROM ${this.table}
                    WHERE profile_id = ${profileId} AND archived = false;
            `,
        };

        try {
            return await this.runQuery(query);
        } catch (err) {
            const message = `Unable to fetch Request(s) with Profile Id ${profileId}`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    };
}
