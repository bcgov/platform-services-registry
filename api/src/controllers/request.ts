//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { HumanActionType, RequestType } from '../db/model/request';
import { AuthenticatedUser } from '../libs/authmware';
import { fulfillRequest } from '../libs/fulfillment';
import { MessageType, sendProvisioningMessage } from '../libs/messaging';
import { archiveProjectSet } from '../libs/profile';
import shared from '../libs/shared';

const dm = new DataManager(shared.pgPool);

export const fetchHumanActionRequests = async (
  { query }: { query: any }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  const { filter } = query;

  try {
    const filterValue = true;
    // Step 1. fetch all requests with matching filter
    const requests = await RequestModel.findActiveByFilter(filter, filterValue);
    // Step 2. return JSON of project Ids
    res.status(200).json(requests);
  } catch (err) {
    const message = `Unable to fetch requests requiring human action`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateRequestHumanAction = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  const { requestId } = params;
  const { type, comment } = body;

  try {
    // Step 1. fetch Request
    const request = await RequestModel.findById(requestId);

    // Step 2. create human_action record
    await RequestModel.createHumanAction({
      requestId: request.id,
      type,
      comment,
      userId: user.id,
    });

    // Step 3.a. If approved: fulfillRequest functionality => create bot_message, send NATS message
    // Step 3.b. if rejected: updateRejectProject => archive ProjectSet, Email PO/TC with comment, complete request;
    if (type === HumanActionType.Approve) {
      await fulfillRequest(request);
      await RequestModel.receivedHumanAction(requestId);
      return res.status(204).end();
    }

    logger.info(`Sending CHES message Project Request Rejected for ${request.profileId}`);
    await sendProvisioningMessage(request.profileId, MessageType.RequestRejected);
    logger.info(`CHES message sent for ${request.profileId}`);

    if ( request.type === RequestType.Create) {
      await archiveProjectSet(request.profileId)
    }

    await RequestModel.updateCompletionStatus(requestId);
    return res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable to update request ${requestId} with human action`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
