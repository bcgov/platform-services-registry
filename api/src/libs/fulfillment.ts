//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import config from '../config';
import { ROLE_IDS } from '../constants';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { ProjectProfile } from '../db/model/profile';
import { Quotas, QuotaSize } from '../db/model/quota';
import { RequestEditType } from '../db/model/request';
import { replaceForDescription } from '../libs/utils';
import { NatsContext, NatsContextAction, NatsContextType, NatsMessage } from '../types';
import { MessageType, sendProvisioningMessage } from './messaging';
import { getQuotaSize } from './profile';
import shared from './shared';

const dm = new DataManager(shared.pgPool);
const { ProfileModel, ContactModel, QuotaModel, NamespaceModel } = dm;

export const fulfillNamespaceProvisioning = async (profileId: number) =>
  new Promise(async (resolve, reject) => {
    try {
      const profile: ProjectProfile = await ProfileModel.findById(profileId);
      const subjectPrefix: string = config.get('nats:subjectPrefix');

      const subject: string = subjectPrefix.concat(profile.primaryClusterName);
      const context: NatsContext = await contextForProvisioning(profileId, false);

      await sendNatsMessage(profileId, {
        natsSubject: subject,
        natsContext: context,
      });

      logger.info(`Sending CHES message (${MessageType.ProvisioningStarted}) for ${profileId}`);
      await sendProvisioningMessage(profileId, MessageType.ProvisioningStarted);
      logger.info(`CHES message sent for ${profileId}`);

      resolve();
    } catch (err) {
      const message = `Unable to fulfill namespace provisioning for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  });

export const fulfillEditRequest = async (profileId: number, requestEditType: RequestEditType, requestEditObject: any): Promise<NatsMessage> =>
  new Promise(async (resolve, reject) => {
    try {
      const profile: ProjectProfile = await ProfileModel.findById(profileId);
      const subjectPrefix: string = config.get('nats:subjectPrefix');

      const subject: string = subjectPrefix.concat(profile.primaryClusterName);
      const context: NatsContext = await contextForEditing(profileId, false, requestEditType, requestEditObject);

      const natsMessage = await sendNatsMessage(profileId, {
        natsSubject: subject,
        natsContext: context,
      });

      resolve(natsMessage);
    } catch (err) {
      const message = `Unable to fulfill namespace provisioning for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  });

export const contextForProvisioning = async (profileId: number, isForSync: boolean): Promise<NatsContext> => {
  try {
    const action = isForSync ? NatsContextAction.Sync : NatsContextAction.Create;
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
    const contacts: Contact[] = await ContactModel.findForProject(profileId);
    const quotaSize: QuotaSize = await getQuotaSize(profile);
    const quotas: Quotas = await QuotaModel.findForQuotaSize(quotaSize);

    return await buildContext(action, profile, contacts, quotaSize, quotas);
  } catch (err) {
    const message = `Unable to create context for provisioning ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

export const contextForEditing = async (profileId: number, isForSync: boolean, requestEditType: RequestEditType, requestEditObject: any): Promise<NatsContext> => {
  try {
    const action = isForSync ? NatsContextAction.Sync : NatsContextAction.Edit;
    let profile: ProjectProfile;
    let quotaSize: QuotaSize;
    let quotas: Quotas;
    let contacts: Contact[];

    if (requestEditType === RequestEditType.ProjectProfile) {
      profile = requestEditObject;
    } else {
      profile = await ProfileModel.findById(profileId);
    }

    if (requestEditType === RequestEditType.QuotaSize) {
      quotaSize = requestEditObject.quota;
      quotas = requestEditObject.quotas;
    } else {
      quotaSize = await getQuotaSize(profile);
      quotas = await QuotaModel.findForQuotaSize(quotaSize);
    }

    if (requestEditType === RequestEditType.Contacts) {
      contacts = requestEditObject;
    } else {
      contacts = await ContactModel.findForProject(profileId);
    }

    return await buildContext(action, profile, contacts, quotaSize, quotas);
  } catch (err) {
    const message = `Unable to create context for updating ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

const buildContext = async (
  action: NatsContextAction, profile: ProjectProfile, contacts: Contact[], quotaSize: QuotaSize, quotas: Quotas
): Promise<NatsContext> => {
  try {
    if (!profile.id) {
      throw new Error('Cant get profile id');
    }
    const namespaces = await NamespaceModel.findForProfile(profile.id);

    const tcContact = contacts.filter(c => c.roleId === ROLE_IDS.TECHNICAL_CONTACT).pop();
    const poContact = contacts.filter(c => c.roleId === ROLE_IDS.PRODUCT_OWNER).pop();

    if (!profile || !tcContact || !poContact || !quotaSize || !quotas || !namespaces) {
      throw new Error('Missing arguments to build nats context');
    }

    return {
      action,
      type: NatsContextType.Standard,
      profileId: profile.id,
      displayName: profile.name,
      newDisplayName: 'NULL',
      description: profile.description,
      quota: quotaSize,
      quotas,
      namespaces,
      technicalContact: {
        userId: tcContact.githubId,
        provider: 'github', // TODO:(JL) Fix as part of #94.
        email: tcContact.email,
      },
      productOwner: {
        userId: poContact.githubId,
        provider: 'github', // TODO:(JL) Fix as part of #94.
        email: poContact.email,
      },
    };
  } catch (err) {
    const message = `Unable to build context for profile ${profile.id}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

const sendNatsMessage = async (profileId: number, natsMessage: NatsMessage): Promise<NatsMessage> => {
  try {
    const nc = shared.nats;
    const { natsSubject, natsContext } = natsMessage;

    nc.on('error', () => {
      const errmsg = `NATS error sending order ${profileId} to ${natsSubject}`;
      throw new Error(errmsg);
    });

    logger.info(`Sending NATS message for ${profileId} to ${natsSubject}`);

    nc.publish(natsSubject, replaceForDescription(natsContext));
    logger.info(`NATS Message sent for ${profileId} to ${natsSubject}`);

    nc.flush(() => {
      nc.removeAllListeners(['error']);
    });

    return natsMessage;
  } catch (err) {
    const message = `Unable to send nats message for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};
