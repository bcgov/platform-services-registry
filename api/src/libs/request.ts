//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { ProjectProfile } from '../db/model/profile';
import { QuotaSize } from '../db/model/quota';
import { Request, RequestEditType, RequestType } from '../db/model/request';
import { AuthenticatedUser } from './authmware';
import { MessageType, sendProvisioningMessage } from './messaging';
import { updateQuotaSize } from './profile';
import shared from './shared';

const dm = new DataManager(shared.pgPool);
const { RequestModel, QuotaModel } = dm;

export const requestProjectProfileCreate = async (profileId: number, user: AuthenticatedUser, requiresHumanAction: boolean): Promise<Request> => {
    try {
        const request = await createRequest(RequestType.Create, user.id, requiresHumanAction, profileId);

        logger.info(`Sending CHES message (${MessageType.ProvisioningStarted}) for ${profileId}`);
        await sendProvisioningMessage(profileId, MessageType.ProvisioningStarted);
        await sendProvisioningMessage(profileId, MessageType.RequestApproval);
        logger.info(`CHES message sent for ${profileId}`);

        return request
    } catch (err) {
        const message = `Unable to create request for profile ${profileId}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

export const requestProjectProfileEdit = async (profileId: number, newProjectProfile: ProjectProfile, user: AuthenticatedUser, requiresHumanAction: boolean = false): Promise<Request> => {
    try {
        const editObject = newProjectProfile;
        return await createRequest(
            RequestType.Edit,
            user.id,
            requiresHumanAction,
            profileId,
            RequestEditType.ProjectProfile,
            editObject
        );
    } catch (err) {
        const message = `Unable to request project-profile edit for profile ${profileId}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

export const processProjectProfileEdit = async (request: Request): Promise<void> => {
    const { ProfileModel } = dm;
    try {
        const newProjectProfile = request.editObject;

        if (!newProjectProfile) {
            return;
        }

        await ProfileModel.update(newProjectProfile.id, newProjectProfile);
    } catch (err) {
        const message = `Unable to process project-profile edit for request ${request.id}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

export const requestProfileContactsEdit = async (profileId: number, newContacts: Contact[], user: AuthenticatedUser, requiresHumanAction: boolean = false): Promise<Request> => {
    try {
        const editObject = newContacts;

        return await createRequest(
            RequestType.Edit,
            user.id,
            requiresHumanAction,
            profileId,
            RequestEditType.Contacts,
            editObject
        );
    } catch (err) {
        const message = `Unable to request contacts edit for profile ${profileId}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

export const processProfileContactsEdit = async (request: Request): Promise<void> => {
    const { ContactModel } = dm;

    try {
        const newContacts: Contact[] = request.editObject;

        const contactPromises = newContacts.map((newContact: Contact) => {
            if (!newContact.id) {
                throw new Error('Cant get contact id');
            }
            return ContactModel.update(newContact.id, newContact);
        });

        await Promise.all(contactPromises);
    } catch (err) {
        const message = `Unable to process profile contacts edit for request ${request.id}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

export const requestProfileQuotaSizeEdit = async (profileId: number, requestedQuotaSize: QuotaSize, user: AuthenticatedUser, requiresHumanAction: boolean = false): Promise<Request> => {
    try {
        const requestType = RequestType.Edit;
        const editType = RequestEditType.QuotaSize;
        const editObject = {
            quota: requestedQuotaSize,
            quotas: await QuotaModel.findForQuotaSize(requestedQuotaSize),
        };

        const request = await createRequest(requestType, user.id, requiresHumanAction, profileId, editType, editObject);

        logger.info(`Sending CHES message Project Edit Notification for ${profileId}`);
        await sendProvisioningMessage(profileId, MessageType.EditRequestStarted);
        await sendProvisioningMessage(profileId, MessageType.RequestApproval);
        logger.info(`CHES message sent for ${profileId}`);

        return request
    } catch (err) {
        const message = `Unable to request quota-size edit for profile ${profileId}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

export const processProfileQuotaSizeEdit = async (request: Request): Promise<void> => {
    const { ProfileModel } = dm;

    try {
        const { profileId, editObject } = request;
        const { quota } = editObject;
        const profile = await ProfileModel.findById(profileId);

        await updateQuotaSize(profile, quota);


        logger.info(`Sending CHES message Project Edit Success for ${profileId}`);
        await sendProvisioningMessage(profileId, MessageType.EditRequestCompleted);
        logger.info(`CHES message sent for ${profileId}`);
    } catch (err) {
        const message = `Unable to process quota-size edit for request ${request.id}`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};

const createRequest = async (
    type: RequestType,userId: number, requiresHumanAction: boolean,
    profileId: number, editType?: RequestEditType, editObject?: any
    ): Promise<Request> => {
    try {
        const existingRequests = await RequestModel.findForProfile(profileId);
        if (existingRequests.length > 0) {
            throw new Error('Cant proceed as the profile has existing request');
        }
        return await RequestModel.create({
            profileId,
            editType,
            editObject,
            type,
            requiresHumanAction,
            isActive: true,
            userId,
        });
    } catch (err) {
        const message = `Unable to process request edit`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};
