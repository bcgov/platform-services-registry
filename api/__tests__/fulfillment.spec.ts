//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { RequestEditType } from '../src/db/model/request';
import { contextForProvisioning, fulfillEditRequest, FulfillmentContextAction, fulfillNamespaceProvisioning } from '../src/libs/fulfillment';

const p1 = path.join(__dirname, 'fixtures/select-profile.json');
const profile = JSON.parse(fs.readFileSync(p1, 'utf8'));

const p2 = path.join(__dirname, 'fixtures/select-profile-contacts.json');
const contacts = JSON.parse(fs.readFileSync(p2, 'utf8'));

const p3 = path.join(__dirname, 'fixtures/select-profile-namespaces.json');
const namespaces = JSON.parse(fs.readFileSync(p3, 'utf8'));

const p4 = path.join(__dirname, 'fixtures/select-quota-small.json');
const quotas = JSON.parse(fs.readFileSync(p4, 'utf8'));
const spec = quotas[0].jsonBuildObject;

jest.mock('../src/libs/profile', () => {
  return {
    getProfileCurrentQuotaSize: jest.fn().mockResolvedValue('small'),
  };
});

describe('Services', () => {

  const client = new Pool().connect();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Provisioning context is created', async () => {

    client.query.mockReturnValueOnce({ rows: profile });
    client.query.mockReturnValueOnce({ rows: contacts });
    client.query.mockReturnValueOnce({ rows: quotas });
    client.query.mockReturnValueOnce({ rows: namespaces });
    const result = await contextForProvisioning(12345, FulfillmentContextAction.Create);

    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
  });

  it('Provisioning context is not created (no contacts)', async () => {

    client.query.mockReturnValueOnce({ rows: profile });
    client.query.mockReturnValueOnce({ rows: [] });
    client.query.mockReturnValueOnce({ rows: quotas });
    client.query.mockReturnValueOnce({ rows: namespaces });

    const result = await contextForProvisioning(12345, FulfillmentContextAction.Create);

    expect(result).not.toBeDefined();
  });

  it('Provisioning context is not created (query fails)', async () => {

    client.query.mockImplementation(() => { throw new Error() });

    await expect(contextForProvisioning(12345, FulfillmentContextAction.Create)).rejects.toThrow();
  });

  it('Namespace provisioning succeeds', async () => {

    client.query.mockReturnValueOnce({ rows: profile });
    client.query.mockReturnValueOnce({ rows: contacts });
    client.query.mockReturnValueOnce({ rows: quotas });
    client.query.mockReturnValueOnce({ rows: namespaces });

    await expect(fulfillNamespaceProvisioning(12345)).resolves.toBeUndefined();
  });

  it('Namespace provisioning fails', async () => {

    client.query.mockReturnValueOnce({ rows: profile });
    client.query.mockReturnValueOnce({ rows: [] });
    client.query.mockReturnValueOnce({ rows: quotas });
    client.query.mockReturnValueOnce({ rows: namespaces });

    await expect(fulfillNamespaceProvisioning(12345)).rejects.toThrow();
  });

  it('Provisioned namespaces updated succeeds', async () => {
    const requestEditObject = {
      quota: 'small',
      quotas: spec,
    };
    const requestEditType = RequestEditType.QuotaSize;

    client.query.mockReturnValueOnce({ rows: profile });
    client.query.mockReturnValueOnce({ rows: contacts });
    client.query.mockReturnValueOnce({ rows: quotas });
    client.query.mockReturnValueOnce({ rows: namespaces });

    await expect(fulfillEditRequest(12345, requestEditType, requestEditObject)).resolves.toBeDefined();
  });
});
