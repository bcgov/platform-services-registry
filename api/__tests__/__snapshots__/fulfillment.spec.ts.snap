// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services Provisioning context is created 1`] = `
Object {
  "action": "create",
  "description": "This is a cool project.",
  "displayName": "Project X",
  "namespaces": Array [
    Object {
      "clusters": Array [
        Object {
          "clusterId": 1,
          "name": "silver",
          "provisioned": false,
        },
      ],
      "name": "4ea35c-tools",
      "namespaceId": 13,
    },
    Object {
      "clusters": Array [
        Object {
          "clusterId": 1,
          "name": "silver",
          "provisioned": false,
        },
      ],
      "name": "4ea35c-test",
      "namespaceId": 14,
    },
    Object {
      "clusters": Array [
        Object {
          "clusterId": 1,
          "name": "silver",
          "provisioned": false,
        },
      ],
      "name": "4ea35c-dev",
      "namespaceId": 15,
    },
    Object {
      "clusters": Array [
        Object {
          "clusterId": 1,
          "name": "silver",
          "provisioned": false,
        },
      ],
      "name": "4ea35c-prod",
      "namespaceId": 16,
    },
  ],
  "newDisplayName": "NULL",
  "productOwner": Object {
    "email": "jane@example.com",
    "provider": "github",
    "userId": "jane1100",
  },
  "profileId": 4,
  "quota": "small",
  "quotas": Object {
    "cpu": Object {
      "limits": 8,
      "requests": 4,
    },
    "memory": Object {
      "limits": "32Gi",
      "requests": "16Gi",
    },
    "storage": Object {
      "backup": "25Gi",
      "block": "50Gi",
      "capacity": "50Gi",
      "file": "50Gi",
    },
  },
  "technicalContact": Object {
    "email": "jim@example.com",
    "provider": "github",
    "userId": "jim1100",
  },
  "type": "standard",
}
`;
